mailbox/plugin/quota-search-scanning/src/main/java/org/apache/james/quota/search/scanning/ClauseConverter.java

```diff
@@ -0,0 +1,103 @@
+ ****************************************************************/
+
+package org.apache.james.quota.search.scanning;
+
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import javax.inject.Inject;
+
+import org.apache.james.core.User;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.Quota;
+import org.apache.james.mailbox.model.QuotaRoot;
+import org.apache.james.mailbox.quota.QuotaCount;
+import org.apache.james.mailbox.quota.QuotaManager;
+import org.apache.james.mailbox.quota.QuotaSize;
+import org.apache.james.mailbox.quota.UserQuotaRootResolver;
+import org.apache.james.quota.search.QuotaClause;
+
+import com.google.common.collect.ImmutableMap;
+
+public class ClauseConverter {
+
+    private final UserQuotaRootResolver quotaRootResolver;
+    private final QuotaManager quotaManager;
+    private final Map<Class<? extends QuotaClause>, Function<QuotaClause, Predicate<User>>> toPredicates;
+
+    @Inject
+    public ClauseConverter(UserQuotaRootResolver quotaRootResolver, QuotaManager quotaManager) {
+        this.quotaRootResolver = quotaRootResolver;
+        this.quotaManager = quotaManager;
+        this.toPredicates = ImmutableMap.of(
+            QuotaClause.And.class, this::andToPredicate,
+            QuotaClause.LessThan.class, this::lessThanToPredicate,
+            QuotaClause.MoreThan.class, this::moreThanToPredicate,
+            QuotaClause.HasDomain.class, this::hasDomainToPredicate);
+    }
+
+    public Predicate<User> andToPredicate(QuotaClause.And and) {
+        return and.getClauses()
+            .stream()
+            .map(this::toPredicate)
+            .reduce((p1, p2) -> (user -> p1.test(user) && p2.test(user)))
+            .orElse(user -> true);
+    }
+
+    public Predicate<User> toPredicate(QuotaClause clause) {
+        return toPredicates.get(clause.getClass())
+            .apply(clause);
+    }
+
+    public Predicate<User> moreThanToPredicate(QuotaClause clause) {
+        QuotaClause.MoreThan moreThan = (QuotaClause.MoreThan) clause;
+        return user -> retrieveUserRatio(user) >= moreThan.getQuotaThreshold().getQuotaOccupationRatio();
+    }
+
+    public Predicate<User> lessThanToPredicate(QuotaClause clause) {
+        QuotaClause.LessThan lessThan = (QuotaClause.LessThan) clause;
+        return user -> retrieveUserRatio(user) <= lessThan.getQuotaThreshold().getQuotaOccupationRatio();
+    }
+
+    public Predicate<User> hasDomainToPredicate(QuotaClause clause) {
+        QuotaClause.HasDomain hasDomain = (QuotaClause.HasDomain) clause;
+        return user -> user.getDomainPart()
+            .map(hasDomain.getDomain()::equals)
+            .orElse(false);
+    }
+
+    public Predicate<User> andToPredicate(QuotaClause clause) {
+        QuotaClause.And and = (QuotaClause.And) clause;
+        return andToPredicate(and);
+    }
+
+    public Double retrieveUserRatio(User user) {
```

Comment:
I think that QuotaManager deserves to have a getUserRatio method

Reply:
Not really... By doing so the QuotaManager would have to be aware that a Quota Root is attached to a user, and Quota manager don't have that knowledge (QuotaRootResolver has).

Reply:
That's true, what about getRootRatio then ?

Reply:
we'll do that later